expan/core/experiment.py

```diff
@@ -1,280 +1,222 @@
         if (len(flags[flags == True]) / float(len(self.data))) > 0.02:
-            warnings.warn('More than 2% of entities have been filtered out, consider adjusting the percentile value.')
-
+            logger.warning('More than 2% of entities have been filtered out, consider adjusting the percentile value.')
         self.data = self.data[flags == False]
 
-    def sga(self, feature_name_to_bins, multi_test_correction=False):
-        """
-        Perform subgroup analysis.
-        Args:
-            feature_name_to_bins (dict): a dict of feature name (key) to list of Bin objects (value). 
-                                      This dict defines how and on which column to perform the subgroup split.
-            multi_test_correction (boolean): flag of whether the correction for multiple testing is needed.
-        Returns:
-            Analysis results per subgroup. 
-        """
 
-        for feature in feature_name_to_bins:
-            # check type
-            if type(feature) is not str:
-                raise TypeError("Key of the input dict needs to be string, indicating the name of dimension.")
-            if type(feature_name_to_bins[feature]) is not list:
-                raise TypeError("Value of the input dict needs to be a list of Bin objects.")
-            # check whether data contains this column
-            if feature not in self.data:
-                raise KeyError("No column %s provided in data." % feature)
-
-        subgroups = []
-        for feature in feature_name_to_bins:
-            for bin in feature_name_to_bins[feature]:
-                subgroup = {'dimension': feature,
-                            'segment': str(bin.representation)}
-                subgroup_data = bin(self.data, feature)
-
-                if not self._isValidForAnalysis(subgroup_data):
-                    continue
-
-                subgroup_res = self._delta(method='fixed_horizon', data=subgroup_data,
-                                           multi_test_correction=multi_test_correction)
-                subgroup['result'] = subgroup_res
-                subgroups.append(subgroup)
-
-        return subgroups
-
-    def _isValidForAnalysis(self, df):
-        """
-        Check whether the quality of data is good enough to perform analysis.
-        Invalid cases can be 1. there is no data
-                             2. the data does not contain all the variants to perform analysis
-        :param df: data in the format of pandas dataframe
-        :return: boolean
+    # ----- below are helper methods ----- #
+    def _is_valid_for_analysis(self, data, test):
+        """ Check whether the quality of data is good enough to perform analysis. Invalid cases can be:
+        1. there is no data
+        2. the data does not contain all the variants to perform analysis
+        
+        :type data: DataFrame
+        :type test: StatisticalTest
+        :rtype: bool 
         """
-        if df is None:
+        if data is None:
```

Comment:
I don't really see a point in checking against `None`. Do we have a good reason to expect a `None` there? Why don't we check  for `data=5` or `data='a string'` then?

Reply:
Because this method is used for subgroup data. After applying ```FeatureFilter``` to the overal data, the data frame could be ```None```. It could be worth checking at least.

Reply:
```
In [25]: df
Out[25]:
   a
0  1
1  2
2  3

In [26]: df[df.a == 4]
Out[26]:
Empty DataFrame
Columns: [a]
Index: []

In [27]: df[df.a == 4] is None
Out[27]: False
```

Reply:
I see. Thanks. Will remove it.